src/providers/tpm_provider/utils.rs

```diff
@@ -84,3 +91,134 @@ pub struct PasswordContext {
     pub context: TpmsContext,
     pub auth_value: Vec<u8>,
 }
+
+pub fn parsec_to_tpm_params(attributes: KeyAttributes) -> Result<KeyParams> {
+    match attributes.key_type {
+        KeyType::RsaKeyPair => {
+            let size = match attributes.key_bits {
+                x @ 1024 | x @ 2048 | x @ 3072 | x @ 4096 => x.try_into().unwrap(), // will not fail on the matched values
+                _ => return Err(ResponseStatus::PsaErrorInvalidArgument),
+            };
+            Ok(KeyParams::Rsa {
+                size,
+                scheme: convert_asym_scheme_to_tpm(attributes.key_policy.key_algorithm)?,
+                pub_exponent: 0,
+            })
+        }
+        KeyType::EccKeyPair { curve_family } => Ok(KeyParams::Ecc {
+            scheme: convert_asym_scheme_to_tpm(attributes.key_policy.key_algorithm)?,
+            curve: convert_curve_to_tpm(curve_family, attributes.key_bits)?,
+        }),
+        _ => Err(ResponseStatus::PsaErrorNotSupported),
+    }
+}
+
+fn convert_asym_scheme_to_tpm(algorithm: Algorithm) -> Result<AsymSchemeUnion> {
+    match algorithm {
+        Algorithm::AsymmetricSignature(AsymmetricSignature::RsaPkcs1v15Sign { hash_alg }) => {
+            Ok(AsymSchemeUnion::RSASSA(convert_hash_to_tpm(hash_alg)?))
+        }
+        Algorithm::AsymmetricSignature(AsymmetricSignature::Ecdsa { hash_alg }) => {
+            Ok(AsymSchemeUnion::ECDSA(convert_hash_to_tpm(hash_alg)?))
+        }
+        _ => Err(ResponseStatus::PsaErrorNotSupported),
+    }
+}
+
+#[allow(deprecated)]
+fn convert_hash_to_tpm(hash: Hash) -> Result<HashingAlgorithm> {
+    match hash {
+        Hash::Sha1 => Ok(HashingAlgorithm::Sha1),
+        Hash::Sha256 => Ok(HashingAlgorithm::Sha256),
+        Hash::Sha384 => Ok(HashingAlgorithm::Sha384),
+        Hash::Sha512 => Ok(HashingAlgorithm::Sha512),
+        Hash::Sha3_256 => Ok(HashingAlgorithm::Sha3_256),
+        Hash::Sha3_384 => Ok(HashingAlgorithm::Sha3_384),
+        Hash::Sha3_512 => Ok(HashingAlgorithm::Sha3_512),
+        _ => Err(ResponseStatus::PsaErrorNotSupported),
+    }
+}
+
+fn convert_curve_to_tpm(curve_family: EccFamily, key_size: u32) -> Result<EllipticCurve> {
+    match (curve_family, key_size) {
+        (EccFamily::SecpR1, 192) => Ok(EllipticCurve::NistP192),
+        (EccFamily::SecpR1, 224) => Ok(EllipticCurve::NistP224),
+        (EccFamily::SecpR1, 256) => Ok(EllipticCurve::NistP256),
+        (EccFamily::SecpR1, 384) => Ok(EllipticCurve::NistP384),
+        (EccFamily::SecpR1, 512) => Ok(EllipticCurve::NistP521),
```

Comment:
Just checking the number, we have 512 on one side and 521 on the other, is it a typo?

Reply:
It's a typo in the spec...